gst/nnstreamer/tensor_query/tensor_query_serversrc.c

```diff
@@ -57,62 +87,265 @@ gst_tensor_query_serversrc_class_init (GstTensorQueryServerSrcClass * klass)
+ */
+static GstFlowReturn
+gst_tensor_query_serversrc_create (GstPushSrc * psrc, GstBuffer ** outbuf)
+{
+  GstTensorQueryServerSrc *src = GST_TENSOR_QUERY_SERVERSRC (psrc);
+  GstMemory *mem = NULL;
+  GstMapInfo map;
+  GstMetaQuery *meta_query;
+  TensorQueryCommandData cmd_data;
+  TensorQueryDataInfo data_info;
+  query_connection_handle conn;
+  guint i;
+
+  if (!src->server_data) {
+    nns_loge ("Server data is NULL");
+    return GST_FLOW_ERROR;
+  }
+
+  while (TRUE) {
+    conn = nnstreamer_query_server_accept (src->server_data);
+    if (!conn) {
+      nns_loge ("Failed to accept connection");
+      goto error;
+    }
+
+    if (nnstreamer_query_receive (conn, &cmd_data, src->timeout) != 0) {
+      nns_logi ("Failed to receive cmd");
+      continue;
+    }
+
+    switch (cmd_data.cmd) {
+      case _TENSOR_QUERY_CMD_REQUEST_INFO:
+        if (gst_tensors_info_is_equal (&cmd_data.data_info.config.info,
+                &src->src_config.info)) {
+          cmd_data.cmd = _TENSOR_QUERY_CMD_RESPOND_APPROVE;
+          /* respond sink config */
+          gst_tensor_query_server_get_sink_config (&cmd_data.data_info.config);
+        } else {
+          /* respond deny with src config */
+          nns_logw ("tensor info is not equal");
+          cmd_data.cmd = _TENSOR_QUERY_CMD_RESPOND_DENY;
+          gst_tensors_config_copy (&cmd_data.data_info.config,
+              &src->src_config);
+        }
+        if (nnstreamer_query_send (conn, &cmd_data, src->timeout) != 0) {
+          nns_logi ("Failed to send respond");
+          continue;
+        }
+        break;
+
+      case _TENSOR_QUERY_CMD_TRANSFER_START:
+        data_info = cmd_data.data_info;
+        *outbuf = gst_buffer_new ();
+        for (i = 0; i < data_info.num_mems; i++) {
+          mem = gst_allocator_alloc (NULL, data_info.mem_sizes[i], NULL);
+          gst_memory_map (mem, &map, G_PARAM_READWRITE);
+          cmd_data.data.data = map.data;
+          cmd_data.data.size = map.size;
+          if (nnstreamer_query_receive (conn, &cmd_data, src->timeout) != 0) {
+            nns_logi ("Failed to receive data");
+            goto reset_buffer;
+          }
+
+          gst_memory_unmap (mem, &map);
+          gst_buffer_append_memory (*outbuf, mem);
+        }
+
+        /* receive end */
+        if (nnstreamer_query_receive (conn, &cmd_data, src->timeout) != 0 ||
+            cmd_data.cmd != _TENSOR_QUERY_CMD_TRANSFER_END) {
+          nns_logi ("Failed to receive end command");
+          goto reset_buffer;
+        }
+
+        meta_query = gst_buffer_add_meta_query (*outbuf);
+        if (meta_query) {
+          meta_query->host =
+              g_strdup (nnstreamer_query_connection_get_host (conn));
+        }
+        return GST_FLOW_OK;
+
+      default:
+        nns_loge ("Invalid cmd type %d", cmd_data.cmd);
+        break;
+    }
+    continue;
+  reset_buffer:
+    gst_allocator_free (NULL, mem);
```

Comment:
Do you need this line when line 326 is executed?

Reply:
I don't think so.... It might cause double free exception? I will fix it. Thank you